packages/@lwc/engine-core/src/framework/hydration.ts

```diff
@@ -243,11 +268,25 @@ function hydrateCustomElement(
     vnode: VCustomElement,
     renderer: RendererAPI
 ): Node | null {
-    if (
-        !hasCorrectNodeType<Element>(vnode, elm, EnvNodeTypes.ELEMENT, renderer) ||
-        !isMatchingElement(vnode, elm, renderer)
-    ) {
-        return handleMismatch(elm, vnode, renderer);
+    const { validationOptOut } = vnode.ctor;
+    const attrsToNotValidate = getAttrsToNotValidate(validationOptOut);
+
+    // The validationOptOut static property can be an array of attribute names.
+    // Any attribute names specified in that array will not be validated, and the
+    // LWC runtime will assume that VDOM attrs and DOM attrs are in sync.
+    //
+    // If validationOptOut is true, no attributes will be checked for correctness
+    // and the runtime will assume VDOM attrs and DOM attrs are in sync.
+    //
+    // Therefore, if validationOptOut is falsey or an array of strings, we need to
+    // examine some or all of the custom element's attributes.
```

Comment:
Rather than check for falsiness, I think it would be simpler to just check for `isTrue` and `isArray`. Anything else can be treated as "no filtering expected."

Otherwise, we may need to add tests for every possible falsey value (undefined, null, 0, empty string, etc.), since historically anything we don't test will eventually break.

Reply:
Took care of this with the refactor related to your above comment.